ui/src/shared/components/CustomTimeRange.js

```diff
@@ -0,0 +1,100 @@
+import React, {PropTypes, Component} from 'react'
+import rome from 'rome'
+import moment from 'moment'
+
+class CustomTimeRange extends Component {
+  constructor(props) {
+    super(props)
+
+    this.handleClick = ::this.handleClick
+    this._formatTimeRange = ::this._formatTimeRange
+  }
+
+  componentDidMount() {
+    const {timeRange} = this.props
+
+    const lower = rome(this.lower, {
+      initialValue: this._formatTimeRange(timeRange.lower),
+    })
+    const upper = rome(this.upper, {
+      initialValue: this._formatTimeRange(timeRange.upper),
+    })
+
+    this.lowerCal = lower
+    this.upperCal = upper
+  }
+
+  // If there is an upper or lower time range set, set the corresponding calendar's value.
+  componentWillReceiveProps(nextProps) {
+    const {lower, upper} = nextProps.timeRange
+    if (lower) {
+      this.lowerCal.setValue(this._formatTimeRange(lower))
+    }
+
+    if (upper) {
+      this.upperCal.setValue(this._formatTimeRange(upper))
+    }
+  }
+
+  render() {
+    return (
+      <div className="custom-time--container">
+        <div className="custom-time--dates">
+          <div className="custom-time--lower" ref={r => (this.lower = r)} />
+          <div className="custom-time--upper" ref={r => (this.upper = r)} />
+        </div>
+        <div
+          className="custom-time--apply btn btn-sm btn-primary"
+          onClick={this.handleClick}
+        >
+          Apply
+        </div>
+      </div>
+    )
+  }
+
+  /*
+   * Upper and lower time ranges are passed in with single quotes as part of
+   * the string literal, i.e. "'2015-09-23T18:00:00.000Z'".  Remove them
+   * before passing the string to be parsed.
+   */
+  _formatTimeRange(timeRange) {
+    if (!timeRange) {
+      return ''
+    }
+
+    // If the given time range is relative, create a fixed timestamp based on its value
+    if (timeRange.match(/^now/)) {
+      const match = timeRange.match(/\d+\w/)[0]
+      const duration = match.slice(0, match.length - 1)
+      const unitOfTime = match[match.length - 1]
+      return moment().subtract(duration, unitOfTime)
```

Comment:
TIL that we support `ms` as a unitOfTime, so we'll have to expand the regex to capture those as well. Apparently we also support `u` and `µ` for "microseconds", but moment doesn't understand those, so if we want to support them, we'll have to transform them.

Additionally I think we can do the `duration`/`unitOfTime` extraction within the regex:

```javascript
if (timeRange.match(/^now/)) {
  [__, duration, unitOfTime] = timeRange.match(/(\d+)(\w+)/)
  moment().subtract(duration, unitOfTime)
}
```